pkg/ec2pricing/ec2pricing.go

```diff
@@ -0,0 +1,312 @@
+}
+
+func (p *EC2Pricing) calculateSpotAggregate(spotPriceEntries []spotPricingEntry) float64 {
+	if len(spotPriceEntries) == 0 {
+		return 0.0
+	}
+	// Sort slice by timestamp in decending order from the end time (most likely, now)
+	sort.Slice(spotPriceEntries, func(i, j int) bool {
+		return spotPriceEntries[i].Timestamp.After(spotPriceEntries[j].Timestamp)
+	})
+
+	endTime := spotPriceEntries[0].Timestamp
+	startTime := spotPriceEntries[len(spotPriceEntries)-1].Timestamp
+	totalDuration := endTime.Sub(startTime).Minutes()
+
+	priceSum := float64(0)
+	for i, entry := range spotPriceEntries {
+		duration := spotPriceEntries[int(math.Max(float64(i-1), 0))].Timestamp.Sub(entry.Timestamp).Minutes()
+		priceSum += duration * entry.SpotPrice
+	}
+	return (priceSum / totalDuration)
+}
+
+// GetOndemandInstanceTypeCost retrieves the on-demand hourly cost for the specified instance type
+func (p *EC2Pricing) GetOndemandInstanceTypeCost(instanceType string) (float64, error) {
+	regionDescription := p.getRegionForPricingAPI()
+	// TODO: mac.metal instances cannot be found with the below filters
+	productInput := pricing.GetProductsInput{
+		ServiceCode: aws.String(serviceCode),
+		Filters: []*pricing.Filter{
+			{Type: aws.String(pricing.FilterTypeTermMatch), Field: aws.String("ServiceCode"), Value: aws.String(serviceCode)},
+			{Type: aws.String(pricing.FilterTypeTermMatch), Field: aws.String("operatingSystem"), Value: aws.String("linux")},
+			{Type: aws.String(pricing.FilterTypeTermMatch), Field: aws.String("location"), Value: aws.String(regionDescription)},
+			{Type: aws.String(pricing.FilterTypeTermMatch), Field: aws.String("capacitystatus"), Value: aws.String("used")},
+			{Type: aws.String(pricing.FilterTypeTermMatch), Field: aws.String("preInstalledSw"), Value: aws.String("NA")},
+			{Type: aws.String(pricing.FilterTypeTermMatch), Field: aws.String("tenancy"), Value: aws.String("shared")},
+			{Type: aws.String(pricing.FilterTypeTermMatch), Field: aws.String("instanceType"), Value: aws.String(instanceType)},
+		},
+	}
+
+	// Check cache first and return it if available
+	if price, ok := p.cache[instanceType]; ok {
+		return price, nil
+	}
+
+	pricePerUnitInUSD := float64(-1)
+	err := p.PricingClient.GetProductsPages(&productInput, func(pricingOutput *pricing.GetProductsOutput, nextPage bool) bool {
+		var err error
+		for _, priceDoc := range pricingOutput.PriceList {
+			_, pricePerUnitInUSD, err = parseOndemandUnitPrice(priceDoc)
+		}
+		if err != nil {
+			// keep going through pages if we can't parse the pricing doc
+			return true
+		}
+		return false
+	})
+	if err != nil {
+		return -1, err
+	}
+	return pricePerUnitInUSD, nil
+}
+
+// HydrateSpotCache makes a bulk request to the spot-pricing-history api to retrieve all instance type pricing and stores them in a local cache
+// If HydrateSpotCache is called more than once, the cache will be fully refreshed
+// There is no TTL on cache entries
```

Comment:
You wouldn't refresh the cache when using the CLI since the cache doesn't persist to disk currently. It might make sense to persist the cache to disk at some point.  Currently the cache is populated on CLI startup, but the only current reason to refresh is if you are using this in an "always-on" system. In that case, you'd want to setup a go routined to periodically refresh the cache. How often you do that is probably app dependent. Prices don't change very often for OD and spot is using an N day avg, so if that N is large, it may not matter to have it be off by days. 

I don't think an S3 cache would help very much. That would basically be the same as getting information from the APIs directly. I think persisting to disk for the CLI might be a good option.